source/agora/cli/gen/main.d

```diff
@@ -0,0 +1,200 @@
+/*******************************************************************************
+
+    Entry point for the agora-gen
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.cli.gen.main;
+
+import scpd.tests.VTableTypes;
+import scpd.tests.VTableTest;
+
+import std.file;
+import std.format;
+import std.stdio;
+import std.traits;
+
+/// Workaround for issue likely related to dub #225,
+/// expects a main() function and invokes it after unittesting.
+version (unittest) void main () { } else:
+
+/*******************************************************************************
+
+    Application entry point
+
+*******************************************************************************/
+
+int main (string[] args)
+{
+    // Validate parameters
+    if (args.length < 2)
+    {
+        writeln("usage: agora-gen [output source file]");
+        return 1;
+    }
+    auto filename = args[1];
+    File file = File(filename, "w+");
+
+    bool first;
+    int method_idx;
+
+    writeHeader(file);
+
+    static foreach (e; VTableCheckClasses)
+    {
+        mixin(
+        q{
+            file.writeln(`///  Uses to test %1$s.`);
+            file.writeln(`class %1$sForTest`);
+            file.writeln(`{`);
+            file.writeln(`    public:`);
+            static foreach (member; __traits(allMembers, %1$s))
+            {
+                mixin(
+                q{
+                    static if (__traits(isVirtualMethod, %1$s.%2$s) && (`%2$s` != `__xdtor`))
+                    {
+                        static if (`%2$s` == `__dtor`)
+                        {
+                            file.writeln(`        virtual ~%1$sForTest () {};`);
+                        }
+                        else
+                        {
+                            file.writeln(`        virtual void %2$s () {};`);
+                        }
+                    }
+                }.format(`%1$s`, member));
+            }
+
+            file.writeln(`};`);
+            file.writeln(``);
+
+
+            file.writeln(`///  Checks the offset of virtual methods inside the class %1$s.`);
+            file.writeln(`int checkVMOffset%1$s ()`);
+            file.writeln(`{`);
+
+            method_idx = 1;
+            static foreach (member; __traits(allMembers, %1$s))
```

Comment:
How are overloads handled ?

Reply:
I don't know why this place needs overloads.
Why do you think it is necessary?
The parameter and return values are not used here.
I knew it was decided not to take care of the override.

Reply:
There is also `getVMOffset...`, what about this?

Reply:
``` D
unittest
{
    static foreach (e; VTableCheckClasses)
    {
        mixin(q{
            assert(checkVMOffset%1$s() == 0, "The virtual method offset of %1$s does not match.");
        }.format(e));
    }
}

```